dozer/cogs/shortcuts.py

```diff
@@ -129,29 +130,51 @@ async def list(self, ctx):
     """
 
     @Cog.listener()
-    async def on_message(self, msg):
+    async def on_message(self, msg: discord.Message):
         """prefix scanner"""
         if not msg.guild or msg.author.bot:
             return
-        setting = await self.settings_cache.query_one(guild_id=msg.guild.id)
-        if setting is None:
-            return
-
-        c = msg.content
-        if len(c) < len(setting.prefix):
-            return
 
-        if not c.startswith(setting.prefix):
+        setting = await ShortcutSetting.get_unique_by(guild_id = msg.guild.id)
+        if not setting:
             return
 
-        shortcuts = await ShortcutEntry.get_by(guild_id=msg.guild.id)
-        if not shortcuts:
-            return
-
-        for shortcut in shortcuts:
-            if c.lower()[len(setting.prefix):] == shortcut.name.lower():
-                await msg.channel.send(shortcut.value)
-                return
+        # Search for the prefix within the message content
+        prefix = setting.prefix
+        prefix_index = msg.content.find(prefix)
+
+        if prefix_index != -1:
+            # before running any chatgpt stuff, check if there's a space between prefix and shortcut
+            if prefix_index + len(prefix) < len(msg.content) and msg.content[prefix_index + len(prefix)] == ' ':
+                return  # there's a space, so it was probably meant to be used in text rather than call a shortcut
+            # Extract the command part from the message directly after the prefix
+            start_index = prefix_index + len(prefix)
+            remaining_content = msg.content[start_index:].strip()
+
+            # Check if the remaining content starts with a valid command name
+            all_shortcuts = await ShortcutEntry.get_by(guild_id = msg.guild.id)
+            all_shortcuts = [s.name for s in all_shortcuts]
+
+            best_match = process.extractOne(remaining_content, all_shortcuts, scorer = fuzz.partial_ratio)
```

Comment:
```python
            best_match = process.extractOne(remaining_content, all_shortcuts, scorer = fuzz.partial_ratio)
```
This is a logic bug, if a message contains e.g.
`gm0 is neat, !roadrumner`, this will find gm0 and say there is no match, because gm0 does not contain a prefix before it.

```py
prefix_index = msg.content.find(prefix)
```
is also a logic bug, as `hi! you should check out !rr` would fail because `find` would find the first `!` (assuming `!` is the prefix.

I would recommend first finding substrings of the form `!(\w|\d|-|_`)+` or so (probably not with regex) (where ! is the server-specific prefix), but honestly this feels like a very large amount of computation to run on every message compared to checking if it starts with a specific substring, for not all that much benefit.

Reply:
So the first one would actually work, as it's only checking after prefix. the second one is a bug though, but then not every server uses ! as their shortcuts prefix, that's mainly a ftc thing. It's also set to not check for shortcuts if there's a space/nothing right after the prefix. If there's more checks for prefixes, however, we might as well just add something to check every instance of the prefix being present, and bring in potential for calling multiple shortcuts in a single message. That, however, opens up a window for increasing computation as it would be possible to send a message with `!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!` or similar as the text. 

That comes down to this: is it worth adding more checks for additional prefixes in the message? Or is checking if there's a shortcut after the first one enough.